test/e2e/storage/external/external.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package external
+
+import (
+	"flag"
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	// "k8s.io/klog"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/testsuites"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// List of testSuites to be executed for each external driver.
+var csiTestSuites = []func() testsuites.TestSuite{
+	testsuites.InitVolumesTestSuite,
+	testsuites.InitVolumeIOTestSuite,
+	testsuites.InitVolumeModeTestSuite,
+	testsuites.InitSubPathTestSuite,
+	testsuites.InitProvisioningTestSuite,
+}
+
+func init() {
+	flag.Var(testDriverParameter{}, "storage.testdriver", "name of a .yaml or .json file that defines a driver for storage testing, can be used more than once")
```

Comment:
How can this be used more than once? Do you just define the flag multiple times in the command line and go just figures it out that you want to init multiple versions of this package? 

It's not clear to me how to invoke this test package, what kind of parameter to pass in to `testDriverParameter` since it seems to be just an empty struct.

I think these are all questions that could be answered in a `README`.

Reply:
maybe another way to do this would be that `testDriverParameter` could point to the file that defines the `DriverInfo`? It seems like maybe thats the idea but i'm not sure why we're using an empty struct that might be implementing an interface (I assume thats where `Set` comes from). It's just a little bit confusing to me

Reply:
Yes, `-storage.testdriver` can be given more than once. Each parameter instantiates the testsuite for another driver. The mechanism for doing that is the `testDriverParameter` and its implementation of the `flag.Value` interface. Let me add the interface check and some comments to make this clearer.

If we just stored the parameter instead of handling it right away, then a) the parameter can only be used once (at least when using a plain string value) and more importantly, b) we have no way  of hooking into the process init between "parsing parameters" and "test suite starts", i.e. we cannot define the tests for that driver (or drivers).

Reply:
Added an explanation.